### ISSUE TITLE
Build error...

### ISSUE THREAD

**Issue by MikeMitterer** [dart-lang/reflectable#61]
I added reflection support to xxgreg's mustache lib: https://github.com/MikeMitterer/mustache (works fine)
and threw out all mirror-references from my MDL/Dart lib: https://goo.gl/XSPuiN (reflections branch)

If I try to compile this sample https://goo.gl/nENSsI to JS I get the error below.

I use 0.4.x because one of the DI dependencies is still using this version...
My main problem is that I don't know which feature you mean - maybe I could avoid it but I have no idea what causes this error. (BTW: everything works fine in VM)

Error:
```
Build error:
Transform Reflectable on mdl_dialog_sample|ReflectableTransformed threw error: *** Unfortunately, this feature has not yet been implemented: Attempt to generate code for an unsupported kind of type (LambdaContext) → Object.
If you wish to ensure that it is prioritized, please report it on github.com/dart-lang/reflectable.
package:reflectable/src/incompleteness.dart 41                unimplementedError
package:reflectable/src/transformer_implementation.dart 1452  _ReflectorDomain._typeCodeOfTypeArgument.fail
package:reflectable/src/transformer_implementation.dart 1479  _ReflectorDomain._typeCodeOfTypeArgument
package:reflectable/src/transformer_implementation.dart 1471  _ReflectorDomain._typeCodeOfTypeArgument.<fn>
dart:_internal/iterable.dart 413                              MappedListIterable.elementAt
dart:_internal/iterable.dart 155                              ListIterable.join
package:reflectable/src/transformer_implementation.dart 1472  _ReflectorDomain._typeCodeOfTypeArgument
package:reflectable/src/transformer_implementation.dart 1503  _ReflectorDomain._typeCodeOfClass
package:reflectable/src/transformer_implementation.dart 894   _ReflectorDomain._generateCode.<fn>.<<fn>_sync_body>.<fn>
dart:_internal/iterable.dart 390                              MappedIterator.moveNext
dart:core-patch/core_patch.dart 54                            _SyncIterator.moveNext
dart:core/iterable.dart 311                                   Iterable.join
package:reflectable/src/transformer_implementation.dart 3551  _formatAsList
package:reflectable/src/transformer_implementation.dart 898   _ReflectorDomain._generateCode
package:reflectable/src/transformer_implementation.dart 127   ReflectionWorld.generateCode.<fn>
dart:_internal/iterable.dart 413                              MappedListIterable.elementAt
dart:_internal/iterable.dart 148                              ListIterable.join
package:reflectable/src/transformer_implementation.dart 3558  _formatAsMap
package:reflectable/src/transformer_implementation.dart 125   ReflectionWorld.generateCode
package:reflectable/src/transformer_implementation.dart 3207  TransformerImplementation._generateNewEntryPoint
package:reflectable/src/transformer_implementation.dart 3350  TransformerImplementation.apply.<async>
dart:async/zone.dart 1149                                     _RootZone.runUnary
dart:async/future_impl.dart 502                               _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 585                               _Future._propagateToListeners
dart:async/future_impl.dart 376                               _Future._completeWithValue
dart:async/future_impl.dart 430                               _Future._asyncComplete.<fn>
dart:async/schedule_microtask.dart 43                         _microtaskLoop
dart:async/schedule_microtask.dart 52                         _microtaskLoopEntry
dart:isolate-patch/isolate_patch.dart 96                      _runPendingImmediateCallback
dart:isolate-patch/isolate_patch.dart 151                     _RawReceivePortImpl._handleMessage

package:reflectable/src/incompleteness.dart 41                unimplementedError
package:reflectable/src/transformer_implementation.dart 1452  _ReflectorDomain._typeCodeOfTypeArgument.fail
package:reflectable/src/transformer_implementation.dart 1479  _ReflectorDomain._typeCodeOfTypeArgument
package:reflectable/src/transformer_implementation.dart 1471  _ReflectorDomain._typeCodeOfTypeArgument.<fn>
dart:_internal                                                ListIterable.join
package:reflectable/src/transformer_implementation.dart 1472  _ReflectorDomain._typeCodeOfTypeArgument
package:reflectable/src/transformer_implementation.dart 1503  _ReflectorDomain._typeCodeOfClass
package:reflectable/src/transformer_implementation.dart 894   _ReflectorDomain._generateCode.<fn>.<<fn>_sync_body>.<fn>
dart:core                                                     Iterable.join
package:reflectable/src/transformer_implementation.dart 3551  _formatAsList
package:reflectable/src/transformer_implementation.dart 898   _ReflectorDomain._generateCode
package:reflectable/src/transformer_implementation.dart 127   ReflectionWorld.generateCode.<fn>
dart:_internal                                                ListIterable.join
package:reflectable/src/transformer_implementation.dart 3558  _formatAsMap
package:reflectable/src/transformer_implementation.dart 125   ReflectionWorld.generateCode
package:reflectable/src/transformer_implementation.dart 3207  TransformerImplementation._generateNewEntryPoint
package:reflectable/src/transformer_implementation.dart 3350  TransformerImplementation.apply.<async>
```

**Comment by MikeMitterer** [comment]
LambdaContext is an abstract class - could this cause the problem? (https://goo.gl/S9Ilwo)

**Comment by eernstg** [comment]
The basic issue is that the situation requires the reflectable transformer to generate an expression which will deliver the type (as an instance of `Type`) of a function. These expressions are typically used to compare with the `runtimeType` of any given object in `Reflectable.reflect`, which creates a connection to reflection on closures, which we don't support yet (it's not easy to get support for that with the currently available primitives).

Could you give some more context about why your software requires some level of reflection on a function type? In general, we have assumed that support for function types would go along with support for reflection on closures, but there may be usages which are more decoupled from closures.

For instance, your need for this `Type` might be coming out of a `typedef` used in a type annotation, along with a `ReflectedTypeCapability`.

We might be able to discover the typedef name and use that in the generated code (but I have no idea about how difficult it would be to do anything reasonably general about this, so I won't promise anything ;-).

**Comment by MikeMitterer** [comment]
Thanks - my problem is that I have no idea which part of my code produces this error. I've commented out all references to xxx.reflect(...) but I still get this error.

I'm using DI in almost all of my samples: https://goo.gl/98OHpF
(This package: https://pub.dartlang.org/packages/di)

Have you heard of any problems with this package in combination with reflections?

**Comment by eernstg** [comment]
I can't immediately check out how you get to reflectable because your 'pubspec.yaml' contains a path dependency (`git clone 'https://github.com/MikeMitterer/dart-material-design-lite-site`, then `grep path: dart-material-design-lite-site/pubspec.yaml`). Can you provide a version of your system which can be inspected (including `pub get` and `pub build`)?

**Comment by MikeMitterer** [comment]
Hi Erik - first of all: Thanks for helping me!
I cloned the mdl/reflections branch into https://github.com/MikeMitterer/dart-material-design-lite-reflections. All the samples in https://github.com/MikeMitterer/dart-material-design-lite-site/tree/reflections/samples are referring this repo. Everything is fine if I run a sample in D-VM but if I pub build it it crashes with the above error.

https://github.com/MikeMitterer/dart-material-design-lite-site/tree/reflections/samples/spa_todo
![screenshot-1959](https://cloud.githubusercontent.com/assets/116654/12302486/b43a465e-ba25-11e5-9134-47dfc592b83f.png)

This sample as TGZ
http://www.mikemitterer.at/fileadmin/_temp_/spa_todo.tgz

**Comment by eernstg** [comment]
First, thanks for reporting issues! ;-)

I tried `pub build --mode=debug test` in the package produced by `git clone https://github.com/MikeMitterer/dart-material-design-lite-reflections`, and it builds without problems (ending in 'Built 3119 files to "build".' which might tell you whether it's complete).

But I can see that the code combines two things that weren't intended to be used together: There is a class `MdlComponentModelAnnotation` which is a subclass of `Reflectable`, so presumably that would be a reflector class, and @MdlComponentModelAnnotation() or some short-hand for that would be used as metadata on classes that require reflection support. That's all fine (it is currently commented out, but I "commented it in" again, and that works nicely).

But the code also contains `MirrorsUsed(..)` on imports of 'dart:mirrors'. A program that uses reflectable should never import 'dart:mirrors', and `@MirrorsUsed(..)` should never be specified. This is because reflectable provides a near-identical service as that offered by 'dart:mirrors', and reflectable allows for transformation which will enable the program to perform reflective operations based on ordinary static (that is, reflection-free) generated code, enabling platforms with no reflection at all, and saving space when the generated code (esp. dart2js) would otherwise be too large.

You can check out https://github.com/dart-lang/reflectable/tree/master/test_reflectable/test to see a number of examples where reflectable is used. The basic idea is that you put `@MdlComponentModelAnnotation()` on each class that needs reflection (there are lots of other options, but that's the most basic one) and specify which *kind* of reflection you need in the class `MdlComponentModelAnnotation` by giving a certain arguments on the form `...Capability` in the super-initializer (which is also what you have already done).

Next, I noted that there are no 'entry_points' and no request for the reflectable transformer in the 'pubspec.yaml'. I added the following to the `transformers:` section:

```yaml
  - reflectable:
      entry_points:
       - test/unit/core/utils_test.dart
       - test/unit/flux/ActionBus_test.dart
       - test/unit/flux/Dispatcher_test.dart
       - test/unit/flux/Emitter_test.dart
       - test/unit/form/form_test.dart
       - test/unit/utils/observables_test.dart
       - test/unit/utils/utils_test.dart
       - test/visual/components/accordion_test.dart
       - test/visual/components/button_test.dart
       - test/visual/components/checkbox_test.dart
       - test/visual/components/data-table_test.dart
       - test/visual/components/icon_toggle_test.dart
       - test/visual/components/layout_test.dart
       - test/visual/components/menu_test.dart
       - test/visual/components/progress_test.dart
       - test/visual/components/radio_test.dart
       - test/visual/core/componenthandler_test.dart
       - test/visual/core/parent_test.dart
       - test/visual/template/components/repeat_test.dart
```

With that, I can reproduce the problem you reported at first (so maybe you had such a 'pubspec.yaml' already at some point, and then changed it ;-).

I've created a patch that will work when applied to reflectable v0.4.0 (that you are currently using). This patch is available at https://codereview.chromium.org/1587113002/, and it makes the `pub build..` process in dart-material-design-lite-reflections succeed.

A similar patch will be applied to the master branch in reflectable such that versions of reflectable (starting with an upcoming 0.5.2) will support generating code for function types as needed here. However, we do not publish a tree of versions, so for v0.4.0 you'll have to fetch the patch and apply it manually. This means that it will be possible-but-inconvenient to handle these function types for a while, but hopefully you'll soon be able to switch to a fresh version of reflectable where this issue has been addressed.

**Comment by eernstg** [comment]
Unfortunately, some function types have empty names, which will cause the patch to create code that does not compile. Currently considering how to work around that.

**Comment by MikeMitterer** [comment]
Cool that you found the issue! It would be no problem to eliminate _DEFAULT_VALUE it's just a helper for debugging. Which function types have empty names? operators? like operator[]...

**Comment by eernstg** [comment]
The empty names were obvious in hindsight: They arise whenever you use C-style syntax to declare a function type for a variable. For instance, `void foo(int aFunction(int _))` declares an argument `aFunction` of type "int -> int", but there is no name for that type.

The approach that we will take at first is to declare that we cannot deliver the corresponding `Type` value. This is not the only case of that kind, e.g., we also cannot deliver the `Type` for `List<E>` inside a class where `E` is a type argument, because we cannot apply a given generic class to a given list of type arguments at runtime. So, with the anonymous function types we'll return false from `hasReflectedType` and `hasDynamicReflectedType`, and then it's up to the programmer to handle it --- possibly by introducing a named function type and using that instead.

**Comment by eernstg** [comment]
The commit https://github.com/dart-lang/reflectable/commit/338e3676ad0843490317cebda45845af20b0c860 is a fix for the issue described here which makes similar changes at the current master branch as those in https://codereview.chromium.org/1587113002/ does for v0.4.0. The update in master is described in https://codereview.chromium.org/1605893002/.

Note that the latter includes more changes, and in particular it adds a test and makes the untransformed code behave in the same manner as the transformed code (such that `typedef`d function types are supported for `reflectedType` and `dynamicReflectedType`, but nameless ones will yield `false` when calling `hasReflectedType`, `hasDynamicReflectedType`). If you need those extras in the context of the old v0.4.0 it should be relatively easy apply the patch from 1605893002 (just patching 'reflectable_mirror_based.dart' would give you the adjustments to the untransformed code behavior).

**Comment by eernstg** [comment]
Closing: the primary issue is resolved and we cannot get closer to full support for function types and closures at this point.

**Comment by MikeMitterer** [comment]
Hi @eernstg, sorry for the late answer but I'm still fighting with "reflectable" I tried to run the tests for this  package: https://github.com/MikeMitterer/dart-mdl-mustache/tree/mdlVersion (! mdlVersion-branch)

Before I use reflectable in my MDL/Dart package the mustache part must run without any problems...
 
If I run pub run test -p "chrome" test/Mustache_test.dart if fails - my problem again: I don't know why it fails! It works in the VM but fails if it is compiled to JS. In this situation the whole thing is almost un-debuggable!!!

This function https://goo.gl/ZFWYFl is responsible for reflecting the vars

I'm really a bit frustrated - two attempts to use the reflectable-package and I failed twice. Transformers are a cool thing in Dart but if a transformer transforms code into something totally different that is not debuggable and results in a different, not predictable (at least not for me)  behavior then it goes in the wrong direction. I thought "reflectable" is the new, official approach for mirrors but it's probably to early to use it for "normal" packages.

```bash
00:46 +92 ~1 -2: test/Mustache_test.dart: Reflectable Simple field
  Value was missing for variable tag: bar. (1:2)
  _{{bar}}_
   ^

  dart:_internal                                    wrapException
  package:mdlmustache/mustache/node.dart 45:35      VariableNode.accept
  package:collection/src/wrappers.dart 514:45       J.x
  package:mdlmustache/mustache/renderer.dart 64:34  Renderer.render.<fn>
  dart:_internal                                    JSArray.1.H.computeSignature.E.func
  package:collection/src/wrappers.dart 514:45       J.ax
  package:mdlmustache/mustache/renderer.dart 64:13  Renderer.render
  package:mdlmustache/mustache/template.dart 61:13  Template.renderString
  test/Mustache_test.dart 657:26                    main.<fn>.<fn>
```

**Comment by floitschG** [comment]
Erik is currently quite busy, but @sigurdm and I had a short look at the problem.
For some reason the transformer isn't run for your test. Since I'm not an expert in `pub run test` I'm CCing @nex3. Hopefully she can help.

In the meantime, we did find a work-around:
- rename the entry-point in pubspec.yaml to be lower case (otherwise it won't match the test file).
- run pub-serve independently (which runs the transformer):
  * `pub serve test`  (in another terminal).
  * `pub run test --pub-server=8080 -p "chrome" test/mustache_test.dart`

Before this works, you have to change the pubspec file slightly. (If you don't `pub run test` gives you the instructions, too). Add the following lines to the transformer section:
```
  - test/pub_serve:
      $include: test/**_test.dart
```

Hopefully, Natalie and Erik can give you the correct way of fixing your problems, but this should unblock you for now.

**Comment by eernstg** [comment]
Hi,

sorry about my response time right now, it'll get better in a few days. One thing that I can contribute is that `pub run test` does not run transformers (as far as I understand there is no bug to report, it just doesn't do that), and https://github.com/dart-lang/test#testing-with-barback explains that the approach mentioned by Florian is the recommended work-around.

I can't immediately see why you'd get the 'Value was missing' message, I would expect an 'Reflectable has not been initialized' before that, but hopefully the situation will be clarified with the transformer in the loop.
